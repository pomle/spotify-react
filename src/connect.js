import { connect as reduxConnect } from 'react-redux';

export function connect(mappers, mapDispatchToProps) {
    function mapStateToProps(state, props) {
        return mappers.reduce((mapStateToProps, mapper) => {
            return Object.assign(mapStateToProps, mapper(state, props));
        }, {});
    }

    return function connect(component) {
        return reduxConnect(mapStateToProps, mapDispatchToProps)(component);
    };
}

export function withPlayingTrack(state) {
    const track = state.player.currentTrack;
    const trackId = track && track.get('id');
    return {
        track,
        analysis: state.track.analysis.get(trackId),
        features: state.track.feature.get(trackId),
        context: state.player.context,
    };
}
