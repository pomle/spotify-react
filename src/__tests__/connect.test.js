import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import { createStore, handleMessage, setAnalysis, setFeature } from '@pomle/spotify-redux';

import { connect, withPlayingTrack } from '../connect';
import playerContext from './fixtures/player-context.json';

function fakeStore(state) {
    return {
        getState() {
            return state;
        },
        dispatch() {

        },
        subscribe() {

        },
    };
}

describe('#connect', () => {
    const FAKE_STORE = fakeStore({
        z: 100,
    });

    it("works as standard connect but takes an array", () => {
        const a = 1, b = 2;

        const Comp = connect([
            state => ({z: state.z, a}),
            state => ({x: state.z + 1, b}),
        ])(({a, b, z, x}) => {
            return <div>{a} {b} {z}, {x}</div>;
        });

        expect(renderer.create(<Comp store={FAKE_STORE}/>))
            .toMatchSnapshot();
    });
});

describe('#withPlayingTrack', () => {
    const store = createStore();
    store.dispatch(handleMessage('state', playerContext));
    store.dispatch(setAnalysis('1Nk3rKuQwAgVNqT93pFvAd', {data: 'The analyses'}));
    store.dispatch(setFeature('1Nk3rKuQwAgVNqT93pFvAd', {data: 'The features'}));

    it("maps Spotify state to props", () => {
        const Comp = connect([
            withPlayingTrack
        ])(props => {
            return <div>
                {props.analysis.data}

                {props.features.data}

                {props.track.get('id')}
            </div>;
        });

        expect(renderer.create(<Comp store={store}/>))
            .toMatchSnapshot();
    });
});
