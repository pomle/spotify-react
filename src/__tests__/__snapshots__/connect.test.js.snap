// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#connect works as standard connect but takes an array 1`] = `
<div>
  1
   
  2
   
  100
  , 
  101
</div>
`;

exports[`#withPlayingTrack maps Spotify state to props 1`] = `
<div>
  The analyses
  The features
  1Nk3rKuQwAgVNqT93pFvAd
</div>
`;
